README.md
Raspi I2C
=========

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/nebrius/raspi-io?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Raspi I2C is part of the [Raspi.js suite](https://github.com/nebrius/raspi) that provides access to the hardware I2C on pins 3 (SDA0) and 5 (SCL0).

If you have a bug report, feature request, or wish to contribute code, please be sure to check out the [Raspi IO Contributing Guide](https://github.com/nebrius/raspi-io/blob/master/CONTRIBUTING.md).

## System Requirements

- Raspberry Pi Model B Rev 1 or newer (sorry Model A users)
- Raspbian Jessie or newer
  - [Node-RED](http://nodered.org/) works, but can be finicky and difficult to debug.
  - See https://github.com/nebrius/raspi-io/issues/24 for more info about support for other OSes
- Node 6.0.0 or newer

Detailed instructions for getting a Raspberry Pi ready for NodeBots, including how to install Node.js, can be found in the [wiki](https://github.com/nebrius/raspi-io/wiki/Getting-a-Raspberry-Pi-ready-for-NodeBots)

## Installation

First, be sure that you have installed [raspi](https://github.com/nebrius/raspi).

Install with npm:

```Shell
npm install raspi-i2c
```

**Note:** this project is written in [TypeScript](http://www.typescriptlang.org/) and includes type definitions in the package.json file. This means that if you want to use it from TypeScript, you don't need to install a separate @types module.

## Example Usage

In TypeScript/ES6:

```TypeScript
import { init } from 'raspi';
import { I2C } from 'raspi-i2c';

init(() => {
  const i2c = new I2C();
  console.log(i2c.readByteSync(0x18)); // Read one byte from the device at address 18
});
```

In JavaScript:

```JavaScript
const raspi = require('raspi');
const I2C = require('raspi-i2c').I2C;

raspi.init(() => {
  const i2c = new I2C();
  console.log(i2c.readByteSync(0x18)); // Read one byte from the device at address 18
});
```

## Notes and Limitations

There are a few limitations and extra steps to be aware of when using I2C on the Raspberry Pi.

First and foremost, be aware that once you use an I2C pin for GPIO, you _cannot_ use it for I2C again until you _reboot_ your Raspberry Pi! If you run the following [Johnny-Five](http://johnny-five.io/) code which leverages this library under the hood, you will get an exception stating "I2C pins not in I2C mode."

```JavaScript
const Raspi = require('raspi-io');
const five = require('johnny-five');
const board = new five.Board({
  io: new Raspi()
});

board.on('ready', () => {
  new five.Pin('SDA');
  board.io.i2cWrite(0x18, 0x5, 'hello');
});
```

Also note that you will need to edit ```/boot/config.txt``` in order to change the I2C baud rate from the default, if you need to. If you notice that behavior is unstable while trying to communicate with another microcontroller, try setting the baudrate to 10000 from the default 100000. This instability has been observed on the Arduino Nano before.

After you install Raspi IO for the first time, you _must_ reboot your Raspberry Pi. I2C support is not enabled by default, and this module runs a script to enable it automatically and adjust a few I2C settings. These settings will not take effect until you reboot your Pi.

Finally, if you try to access a device that doesn't exist, you will get an error stating ```EIO, i/o error``` (sorry it's not very descriptive).

## API

### new I2C(pins)

Instantiates a new I2C instance on the given pins. Note that I2C is limited to only 1 pair of pins on all current models of the Raspberry Pi. Check the [wiring information wiki](https://github.com/nebrius/raspi-io/wiki) for more information.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>pins (optional)</td>
    <td>Number[] | Object</td>
    <td>The configuration of the I2C port. If the config is an array of numbers or strings, it is assumed to be the SDA and SCL pins for the peripheral. If it is an object, the following properties are supported:</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Property</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>pins</td>
          <td></td>
          <td></td>
        </tr>
      </table>
    </td>
  </tr>
</table>

### Instance Methods

#### read(address, register, length, cb)

Reads data from the peripheral at _address_ from the optional _register_ and calls the callback once _length_ bytes have been read.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>length</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be between 0 and 0xff.</td>
  </tr>
  <tr>
    <td>cb</td>
    <td>Function</td>
    <td>The callback to call once the read is complete</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Argument</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>err</td>
          <td>String | null</td>
          <td>The error, if one occurred, else null</td>
        </tr>
        <tr>
          <td>data</td>
          <td>null | Buffer</td>
          <td>If no error occurred, the read data</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

_Returns_: None

#### readSync(address, register, length)

Synchronous version of ```read```.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>length</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be between 0 and 0xff.</td>
  </tr>
</table>

_Returns_: A buffer containing the read data

#### readByte(address, register, cb)

Reads a single byte from the peripheral at _address_ from the optional _register_ and calls the callback.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>cb</td>
    <td>Function</td>
    <td>The callback to call once the read is complete</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Argument</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>err</td>
          <td>String | null</td>
          <td>The error, if one occurred, else null</td>
        </tr>
        <tr>
          <td>data</td>
          <td>null | Number</td>
          <td>If no error occurred, the read byte</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

_Returns_: None

#### readByteSync(address, register)

Synchronous version of ```readByte```.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
</table>

_Returns_: The byte in the form of a number.

#### readWord(address, register, cb)

Reads a single word from the peripheral at _address_ from the optional _register_ and calls the callback.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>cb</td>
    <td>Function</td>
    <td>The callback to call once the read is complete</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Argument</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>err</td>
          <td>String | null</td>
          <td>The error, if one occurred, else null</td>
        </tr>
        <tr>
          <td>data</td>
          <td>null | Number</td>
          <td>If no error occurred, the read byte</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

_Returns_: None

#### readWordSync(address, register)

Synchronous version of ```readWord```.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
</table>

_Returns_: The word in the form of a number.

#### write(address, register, buffer, cb)

Writes the data in _buffer_ to the peripheral at _address_ to the optional _register_ and calls the callback once the write has completed.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>buffer</td>
    <td>Buffer</td>
    <td>The data to write to the peripheral.</td>
  </tr>
  <tr>
    <td>cb</td>
    <td>Function</td>
    <td>The callback to call once the write is complete</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Argument</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>err</td>
          <td>String | null</td>
          <td>The error, if one occurred, else null</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

_Returns_: None

#### writeSync(address, register, buffer)

Synchronous version of ```write```.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to read from. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>buffer</td>
    <td>Buffer</td>
    <td>The data to write to the peripheral.</td>
  </tr>
</table>

_Returns_: None

#### writeByte(address, register, byte, cb)

Writes _byte_ to peripheral at _address_ to the optional _register_ and calls the callback.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to write to. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>byte</td>
    <td>Number</td>
    <td>The byte to write. The value is coerced to an unsigned 8-bit integer.</td>
  </tr>
  <tr>
    <td>cb</td>
    <td>Function</td>
    <td>The callback to call once the write is complete</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Argument</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>err</td>
          <td>String | null</td>
          <td>The error, if one occurred, else null</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

_Returns_: None

#### writeByteSync(address, register, byte)

Synchronous version of ```writeByte```.

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to write to. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>byte</td>
    <td>Number</td>
    <td>The byte to write. The value is coerced to an unsigned 8-bit integer.</td>
  </tr>
</table>

_Returns_: None

#### writeWord(address, register, word, cb)

Writes _word_ to peripheral at _address_ to the optional _register_ and calls the callback.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to write to. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>word</td>
    <td>Number</td>
    <td>The word to write. The value is coerced to an unsigned 16-bit integer.</td>
  </tr>
  <tr>
    <td>cb</td>
    <td>Function</td>
    <td>The callback to call once the write is complete</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Argument</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>err</td>
          <td>String | null</td>
          <td>The error, if one occurred, else null</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

_Returns_: None

#### writeWordSync(address, register, word)

Synchronous version of ```writeWord```.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>address</td>
    <td>Number</td>
    <td>The address of the peripheral to read from. Must be 0 or greater and less than 0x80.</td>
  </tr>
  <tr>
    <td>register (optional)</td>
    <td>Number</td>
    <td>The register to write to. Must be between 0 and 0x7f.</td>
  </tr>
  <tr>
    <td>word</td>
    <td>Number</td>
    <td>The word to write. The value is coerced to an unsigned 16-bit integer.</td>
  </tr>
</table>

_Returns_: None

License
=======

The MIT License (MIT)

Copyright (c) 2014-2017 Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in

CHANGELOG.md
## 6.2.1 (2019-1-10)

- Got rid of duplicate callback TypeScript types and tightened up the code a bit as a result. No behavior change outside of TypeScript.

## 6.2.0 (2018-12-10)

- Updated TypeScript to v3. Note: the compiled output changed as I previously didn't set the compiled output target flag correctly.
- Added `module` export to conform with Core IO types
- Updated to i2c-bus v4.
    - Note: there are breaking changes in i2c-bus 4, but they are not surfaced in raspi-i2c.

## 6.1.2 (2018-4-21)

- Updated to i2c-bus v3.
    - Note: there are breaking changes in i2c-bus v3, but they are not surfaced in raspi-i2c.

## 6.1.1 (2018-1-16)

- Fixed a bug where readWord wasn't properly swizzling the registerOrCb argument as a register when all parameters supplied

## 6.1.0 (2017-9-15)

- Removed logic for parsing pin information passed to the constructor
    - Normally this would be a breaking change, but it was already broken and the current behavior does nothing anyways, meaning there isn't _actually_ a change in behavior. Funny that.
    - Eventually, I want to rework this to provide support for multiple busses.

## 6.0.0 (2017-9-12)

- POTENTIALLY BREAKING CHANGE: renamed the following private properties to begin with an underscore because I just learned TypeScript doesn't do anything to hide private properties. Don't use these properties, changes may not be semver-major next time.
    - `I2C.devices` -> `I2C._devices`
    - `I2C.getDevice` -> `I2C._getDevice`
- BREAKING CHANGE: dropped support for Node.js < 4.0.0, and now enforce it via package.json "engines" field.
- SORT OF BREAKING-ISH CHANGE: dropped support for attempting to install on non-arm platforms via package.json's "cpu" field.
    - Attempting to install this on a non-Raspberry Pi platform before gave a bunch of obtuse errors, so this doesn't _actually_ change the ability to install raspi-i2c, but does make it fail earlier and harder.

## 5.0.4 (2017-5-20)

- Fixed a bug in `writeSync` where we weren't properly swizzling arguments.

## 5.0.3 (2017-22-1)

- Publishing a new version to update the README on npmjs.com. No other changes.

## 5.0.2 (2016-12-30)

- Added some important documentation to the README that needs to be on npmjs.com
  - Note: there are no code changes, only documentation

## 5.0.1 (2016-12-21)

- Added missing TypeScript definition package.json entry

## 5.0.0 (2016-12-21)

- Made all write callbacks optional
- Normalized read and write callbacks so that error is always null if no error occured
  - POTENTIALLY BREAKING CHANGE: If you were using `err === undefined` or something similar, this will no longer work
- Normalized write callbacks so they only take an error argument
  - POTENTIALLY BREAKING CHANCE: If you were relying on the undocumented extra parameters to the write callback, these will now be `undefined`
- Converted the project to TypeScript and cleaned up a bunch of odds and end

## 4.0.0 (2016-11-23)

- Removed 32 byte length limitation when reading/writing directly to the bus and not to a register
  - POTENTIALLY BREAKING CHANGE: if you were relying on the exception being thrown in this situation, you'll need to update your code.

## 3.1.1 (2016-10-29)

- Updated Dependency

## 3.1.0 (2016-7-7)

- Switched dependency ranges to ^
- Bumped dependencies to bring in support for a new Raspberry Pi Zero revision

## 3.0.0 (2016-3-20)

- Dependency update to fix bug
- New build system
- Removed execSync and Symbol shims
  - BREAKING CHANGE: Node 0.10 is no longer supported as a result

## 2.3.2 (2016-3-7)

- Dependency update to add missing Raspberry Pi 3 Model B revision

## 2.3.1 (2016-3-4)

- Forgot to bump the raspi-peripheral dependency in 2.3.0

## 2.3.0 (2016-3-4)

- Added support for the Raspberry Pi 3 Model B

## 2.2.0 (2016-1-6)

- Added support for Ubuntu

## 2.1.0 (2015-12-8)

- Updated dependencies to add Raspberry Pi Zero support

## 2.0.0 (2015-10-21)

- Upgraded i2c-bus to new version that uses NAN 2
  - POTENTIAL BREAKING CHANGE
  - The API has not changed, but the build requirements have
  - Make sure you are running Raspbian Jessie because this module no longer builds on stock Raspbian Wheezy
  - See https://github.com/fivdi/onoff/wiki/Node.js-v4-and-native-addons for more information

## 1.1.0 (2015-10-12)

- Dependency updates to fix bug with invalid pin aliases
- Updated build dependencies

## 1.0.6 (2015-9-3)

- Dependency updates to fix a bug with pin aliasing

## 1.0.5 (2015-8-10)

- Added a Symbol shim to fix a crash in Node.js 0.10

## 1.0.4 (2015-7-16)

- Updated dependencies
- Updated the repository links to point to their new location
- Added a contributing guide
- Added code linter
- Update code style to use newer best practices

## 1.0.3 (2015-5-26)

- Added a default baudrate value to the install script
- Added notes about the default clock rate being unstable with certain Arduinos

## 1.0.2 (2015-5-4)

- Added documentation
- Updated dependencies and reworked code to use ```validateAlive``` internally.
  - No user-facing changes or bug fixes.

## 1.0.0 (2015-3-23)

- First publish


src/index.ts
/*
The MIT License (MIT)

Copyright (c) Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

import { openSync, I2cBus, BufferCallback, ResultCallback } from 'i2c-bus';
import { execSync } from 'child_process';
import { Peripheral } from 'raspi-peripheral';
import { VERSION_1_MODEL_B_REV_1, getBoardRevision } from 'raspi-board';
import { II2C, II2CModule, I2CReadBufferCallback, I2CReadNumberCallback, I2CWriteCallback } from 'core-io-types';

function checkAddress(address: any) {
  if (typeof address !== 'number' || address < 0 || address > 0x7f) {
    throw new Error(`Invalid I2C address ${address}. Valid addresses are 0 through 0x7f.`);
  }
}

function checkRegister(register: any) {
  if (register !== undefined &&
      (typeof register !== 'number' || register < 0 || register > 0xff)
  ) {
    throw new Error(`Invalid I2C register ${register}. Valid registers are 0 through 0xff.`);
  }
}

function checkLength(length: any, hasRegister: boolean) {
  if (typeof length !== 'number' || length < 0 || (hasRegister && length > 32)) {
    // Enforce 32 byte length limit only for SMBus.
    throw new Error(`Invalid I2C length ${length}. Valid lengths are 0 through 32.`);
  }
}

function checkBuffer(buffer: any, hasRegister: boolean) {
  if (!Buffer.isBuffer(buffer) || buffer.length < 0 || (hasRegister && buffer.length > 32)) {
    // Enforce 32 byte length limit only for SMBus.
    throw new Error(`Invalid I2C buffer. Valid lengths are 0 through 32.`);
  }
}

function checkByte(byte: any) {
  if (typeof byte !== 'number' || byte < 0 || byte > 0xff) {
    throw new Error(`Invalid I2C byte ${byte}. Valid values are 0 through 0xff.`);
  }
}

function checkWord(word: any) {
  if (typeof word !== 'number' || word < 0 || word > 0xffff) {
    throw new Error(`Invalid I2C word ${word}. Valid values are 0 through 0xffff.`);
  }
}

function checkCallback(cb: any) {
  if (typeof cb !== 'function') {
    throw new Error('Invalid I2C callback');
  }
}

function createReadBufferCallback(suppliedCallback?: I2CReadBufferCallback): BufferCallback {
  return (err: any, resultOrBytesRead?: Buffer | number, result?: Buffer) => {
    if (suppliedCallback) {
      if (err) {
        suppliedCallback(err, null);
      } else if (typeof result !== 'undefined') {
        suppliedCallback(null, result);
      } else if (Buffer.isBuffer(resultOrBytesRead)) {
        suppliedCallback(null, resultOrBytesRead);
      }
    }
  };
}

function createReadNumberCallback(suppliedCallback?: I2CReadNumberCallback): ResultCallback<number> {
  return (err: any, result?: number) => {
    if (suppliedCallback) {
      if (err) {
        suppliedCallback(err, null);
      } else if (typeof result !== 'undefined') {
        suppliedCallback(null, result);
      }
    }
  };
}

function createWriteCallback(suppliedCallback?: I2CWriteCallback): any {
  return (err: any) => {
    if (suppliedCallback) {
      suppliedCallback(err || null);
    }
  };
}

export class I2C extends Peripheral implements II2C {

  private _devices: I2cBus[] = [];

  constructor() {
    super([ 'SDA0', 'SCL0' ]);
    execSync('modprobe i2c-dev');
  }

  public destroy() {
    this._devices.forEach((device) => device.closeSync());
    this._devices = [];
    super.destroy();
  }

  public read(address: number, length: number, cb: I2CReadBufferCallback): void;
  public read(address: number, register: number, length: number, cb: I2CReadBufferCallback): void;
  public read(
    address: number,
    registerOrLength: number,
    lengthOrCb: number | I2CReadBufferCallback,
    cb?: I2CReadBufferCallback
  ): void {
    this.validateAlive();

    let length: number;
    let register: number | undefined;
    if (typeof cb === 'function' && typeof lengthOrCb === 'number') {
      length = lengthOrCb;
      register = registerOrLength;
    } else if (typeof lengthOrCb === 'function') {
      cb = lengthOrCb;
      length = registerOrLength;
      register = undefined;
    } else {
      throw new TypeError('Invalid I2C read arguments');
    }

    checkAddress(address);
    checkRegister(register);
    checkLength(length, !!register);
    checkCallback(cb);

    const buffer = new Buffer(length);

    if (register === undefined) {
      this._getDevice(address).i2cRead(address, length, buffer, createReadBufferCallback(cb));
    } else {
      this._getDevice(address).readI2cBlock(address, register, length, buffer, createReadBufferCallback(cb));
    }
  }

  public readSync(address: number, registerOrLength: number | undefined, length?: number): Buffer {
    this.validateAlive();

    let register: number | undefined;
    if (typeof length === 'undefined') {
      length = +(registerOrLength as number);
    } else {
      register = registerOrLength;
      length = +length;
    }

    checkAddress(address);
    checkRegister(register);
    checkLength(length, !!register);

    const buffer = new Buffer(length);

    if (register === undefined) {
      this._getDevice(address).i2cReadSync(address, length, buffer);
    } else {
      this._getDevice(address).readI2cBlockSync(address, register, length, buffer);
    }

    return buffer;
  }

  public readByte(address: number, cb: I2CReadNumberCallback): void;
  public readByte(address: number, register: number, cb: I2CReadNumberCallback): void;
  public readByte(address: number, registerOrCb: number | I2CReadNumberCallback, cb?: I2CReadNumberCallback): void {
    this.validateAlive();

    let register: number | undefined;
    if (typeof registerOrCb === 'function') {
      cb = registerOrCb;
      register = undefined;
    }

    checkAddress(address);
    checkRegister(register);
    checkCallback(cb);

    if (register === undefined) {
      const buffer = new Buffer(1);
      this._getDevice(address).i2cRead(address, buffer.length, buffer, (err) => {
        if (err) {
          if (cb) {
            cb(err, null);
          }
        } else if (cb) {
          cb(null, buffer[0]);
        }
      });
    } else {
      this._getDevice(address).readByte(address, register, createReadNumberCallback(cb));
    }
  }

  public readByteSync(address: number, register?: number) {
    this.validateAlive();

    checkAddress(address);
    checkRegister(register);

    let byte: number;
    if (register === undefined) {
      const buffer = new Buffer(1);
      this._getDevice(address).i2cReadSync(address, buffer.length, buffer);
      byte = buffer[0];
    } else {
      byte = this._getDevice(address).readByteSync(address, register);
    }
    return byte;
  }

  public readWord(address: number, cb: I2CReadNumberCallback): void;
  public readWord(address: number, register: number, cb: I2CReadNumberCallback): void;
  public readWord(address: number, registerOrCb: number | I2CReadNumberCallback, cb?: I2CReadNumberCallback): void {
    this.validateAlive();

    let register: number | undefined;
    if (typeof registerOrCb === 'function') {
      cb = registerOrCb;
    } else {
      register = registerOrCb;
    }

    checkAddress(address);
    checkRegister(register);
    checkCallback(cb);

    if (register === undefined) {
      const buffer = new Buffer(2);
      this._getDevice(address).i2cRead(address, buffer.length, buffer, (err) => {
        if (cb) {
          if (err) {
            return cb(err, null);
          }
          cb(null, buffer.readUInt16LE(0));
        }
      });
    } else {
      this._getDevice(address).readWord(address, register, createReadNumberCallback(cb));
    }
  }

  public readWordSync(address: number, register?: number): number {
    this.validateAlive();

    checkAddress(address);
    checkRegister(register);

    let byte: number;
    if (register === undefined) {
      const buffer = new Buffer(2);
      this._getDevice(address).i2cReadSync(address, buffer.length, buffer);
      byte = buffer.readUInt16LE(0);
    } else {
      byte = this._getDevice(address).readWordSync(address, register);
    }
    return byte;
  }

  public write(address: number, buffer: Buffer, cb?: I2CWriteCallback): void;
  public write(address: number, register: number, buffer: Buffer, cb?: I2CWriteCallback): void;
  public write(
    address: number,
    registerOrBuffer: number | Buffer,
    bufferOrCb?: Buffer | I2CWriteCallback,
    cb?: I2CWriteCallback
  ): void {
    this.validateAlive();

    let buffer: Buffer;
    let register: number | undefined;
    if (Buffer.isBuffer(registerOrBuffer)) {
      cb = bufferOrCb as I2CWriteCallback;
      buffer = registerOrBuffer;
      register = undefined;
    } else if (typeof registerOrBuffer === 'number' && Buffer.isBuffer(bufferOrCb)) {
      register = registerOrBuffer;
      buffer = bufferOrCb;
    } else {
      throw new TypeError('Invalid I2C write arguments');
    }

    checkAddress(address);
    checkRegister(register);
    checkBuffer(buffer, !!register);

    if (register === undefined) {
      this._getDevice(address).i2cWrite(address, buffer.length, buffer, createWriteCallback(cb));
    } else {
      this._getDevice(address).writeI2cBlock(address, register, buffer.length, buffer, createWriteCallback(cb));
    }
  }

  public writeSync(address: number, buffer: Buffer): void;
  public writeSync(address: number, register: number, buffer: Buffer): void;
  public writeSync(address: number, registerOrBuffer: number | Buffer, buffer?: Buffer): void {
    this.validateAlive();

    let register: number | undefined;
    if (Buffer.isBuffer(registerOrBuffer)) {
      buffer = registerOrBuffer;
    } else {
      if (!buffer) {
        throw new Error('Invalid I2C write arguments');
      }
      register = registerOrBuffer;
    }

    checkAddress(address);
    checkRegister(register);
    checkBuffer(buffer, !!register);

    if (register === undefined) {
      this._getDevice(address).i2cWriteSync(address, buffer.length, buffer);
    } else {
      this._getDevice(address).writeI2cBlockSync(address, register, buffer.length, buffer);
    }
  }

  public writeByte(address: number, byte: number, cb?: I2CWriteCallback): void;
  public writeByte(address: number, register: number, byte: number, cb?: I2CWriteCallback): void;
  public writeByte(
    address: number,
    registerOrByte: number,
    byteOrCb?: number | I2CWriteCallback,
    cb?: I2CWriteCallback
  ): void {
    this.validateAlive();

    let byte: number;
    let register: number | undefined;
    if (typeof byteOrCb === 'number') {
      byte = byteOrCb;
      register = registerOrByte;
    } else {
      cb = byteOrCb;
      byte = registerOrByte;
    }

    checkAddress(address);
    checkRegister(register);
    checkByte(byte);

    if (register === undefined) {
      this._getDevice(address).i2cWrite(address, 1, new Buffer([byte]), createWriteCallback(cb));
    } else {
      this._getDevice(address).writeByte(address, register, byte, createWriteCallback(cb));
    }
  }

  public writeByteSync(address: number, registerOrByte: number, byte?: number): void {
    this.validateAlive();

    let register: number | undefined;
    if (byte === undefined) {
      byte = registerOrByte;
    } else {
      register = registerOrByte;
    }

    checkAddress(address);
    checkRegister(register);
    checkByte(byte);

    if (register === undefined) {
      this._getDevice(address).i2cWriteSync(address, 1, new Buffer([byte]));
    } else {
      this._getDevice(address).writeByteSync(address, register, byte);
    }
  }

  public writeWord(address: number, word: number, cb?: I2CWriteCallback): void;
  public writeWord(address: number, register: number, word: number, cb?: I2CWriteCallback): void;
  public writeWord(
    address: number,
    registerOrWord: number,
    wordOrCb?: number | I2CWriteCallback,
    cb?: I2CWriteCallback
  ): void {
    this.validateAlive();

    let register: number | undefined;
    let word: number;
    if (typeof wordOrCb === 'number') {
      register = registerOrWord;
      word = wordOrCb;
    } else if (typeof wordOrCb === 'function') {
      word = registerOrWord;
      cb = wordOrCb;
    } else {
      throw new Error('Invalid I2C write arguments');
    }

    checkAddress(address);
    checkRegister(register);
    checkWord(word);

    if (register === undefined) {
      const buffer = new Buffer(2);
      buffer.writeUInt16LE(word, 0);
      this._getDevice(address).i2cWrite(address, buffer.length, buffer, createWriteCallback(cb));
    } else {
      this._getDevice(address).writeWord(address, register, word, createWriteCallback(cb));
    }
  }

  public writeWordSync(address: number, registerOrWord: number, word?: number): void {
    this.validateAlive();

    let register: number | undefined;
    if (word === undefined) {
      word = registerOrWord;
    } else {
      register = registerOrWord;
    }

    checkAddress(address);
    checkRegister(register);
    checkWord(word);

    if (register === undefined) {
      const buffer = new Buffer(2);
      buffer.writeUInt16LE(word, 0);
      this._getDevice(address).i2cWriteSync(address, buffer.length, buffer);
    } else {
      this._getDevice(address).writeWordSync(address, register, word);
    }
  }

  private _getDevice(address: number) {
    let device = this._devices[address];

    if (device === undefined) {
      device = openSync(getBoardRevision() === VERSION_1_MODEL_B_REV_1 ? 0 : 1);
      this._devices[address] = device;
    }

    return device;
  }
}

export const module: II2CModule = {
  createI2C() {
    return new I2C();
  }
};
